process/block/alteredaccounts/tokensProcessor.go

```diff
@@ -0,0 +1,164 @@
+			core.BuiltInFunctionESDTNFTAddQuantity:   {},
+			core.BuiltInFunctionESDTNFTCreate:        {},
+			core.BuiltInFunctionMultiESDTNFTTransfer: {},
+		},
+		shardCoordinator: shardCoordinator,
+	}
+}
+
+func (tp *tokensProcessor) extractESDTAccounts(
+	txPool *indexer.Pool,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) error {
+	var err error
+	for _, txLog := range txPool.Logs {
+		for _, event := range txLog.LogHandler.GetLogEvents() {
+			err = tp.processEvent(event, markedAlteredAccounts)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (tp *tokensProcessor) processEvent(
+	event data.EventHandler,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) error {
+	_, isEsdtOperation := tp.fungibleTokensIdentifiers[string(event.GetIdentifier())]
+	if isEsdtOperation {
+		err := tp.extractEsdtData(event, zeroBigInt, markedAlteredAccounts)
+		if err != nil {
+			log.Debug("cannot extract esdt data", "error", err)
+			return err
+		}
+
+		return nil
+	}
+
+	_, isNftOperation := tp.nonFungibleTokensIdentifier[string(event.GetIdentifier())]
+	if isNftOperation {
+		topics := event.GetTopics()
+		if len(topics) == 0 {
+			return nil
+		}
+
+		nonce := topics[idxTokenNonceInTopics]
+		nonceBigInt := big.NewInt(0).SetBytes(nonce)
+		err := tp.extractEsdtData(event, nonceBigInt, markedAlteredAccounts)
+		if err != nil {
+			log.Debug("cannot extract nft data", "error", err)
+			return nil
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (tp *tokensProcessor) extractEsdtData(
+	event data.EventHandler,
+	nonce *big.Int,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) error {
+	address := event.GetAddress()
+	topics := event.GetTopics()
+	if len(topics) == 0 {
+		return nil
+	}
+
+	// in case of esdt, nft or multi esdt transfers, the 3rd index of the topics contains the destination address
+	tokenID := topics[idxTokenIDInTopics]
+	err := tp.processEsdtDataForAddress(address, nonce, string(tokenID), markedAlteredAccounts)
+	if err != nil {
+		return err
+	}
+
+	if len(topics) > idxReceiverAddressInTopics {
+		destinationAddress := topics[idxReceiverAddressInTopics]
+		err = tp.processEsdtDataForAddress(destinationAddress, nonce, string(tokenID), markedAlteredAccounts)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tp *tokensProcessor) processEsdtDataForAddress(
+	address []byte,
+	nonce *big.Int,
+	tokenID string,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) error {
+	if !tp.isSameShard(address) {
+		return nil
+	}
+
+	addressStr := string(address)
+	markedAccount, exists := markedAlteredAccounts[addressStr]
+	if !exists {
+		markedAccount = &markedAlteredAccount{}
+		markedAlteredAccounts[addressStr] = markedAccount
+	}
+
+	if markedAccount.tokens == nil {
+		markedAccount.tokens = make(map[string]*markedAlteredAccountToken)
+	}
+
+	tokenKey := tokenID + string(nonce.Bytes())
+	_, alreadyExists := markedAccount.tokens[tokenKey]
```

Comment:
[just for my clarification]
If you have, let's say 3 logs with same address = address1, same esdt token = esdt1 => each of them will be called here with nonce =0 => we will only store it once (becase `alreadyExists==true`). Doesn't that mean that we somehow "loose" some intermediary possible balance changes, since you only get the balance at the end of the execution?

Reply:
If it is desired, then perhaps add some unit tests where:
- You have the same esdt token(let's say 2-3 times) for an account => you **only store it once with nonce = 0**
- You have the same nft token(for let's say 2-3 times) for an account => you store **the same nft multiple times with different nonces**

Reply:
As discussed, I am only interested in fetching the data at the end of the block. This way I am sure the data is accurate, regardless of the operations and their order.

Regarding tests:
1. added an additional check on `testExtractAlteredAccountsFromPoolShouldIncludeESDT` to ensure 2 entries won't be added
2. we already have `testExtractAlteredAccountsFromPoolAddressHasMultipleNfts`

Reply:
I understand now. Yes, good catch. done